### ISSUE TITLE
[accname] Two necessary steps for using aria-label values in accessible-name computation are missing from the spec

### ISSUE THREAD

**Issue by sideshowbarker** [w3c/accname#247]
https://w3c.github.io/accname/#comp_labelledby

I have found that:

1. The _“div group explicitly labelledby self and heading”_ subtest at https://wpt.fyi/results/accname/name/comp_labelledby.html won’t pass unless when performing the “For each IDREF” substep of the “LabelledBy” substep in the accessible-name computation algorithm, I check for any `aria-label` attribute and use its value if found.
2.  The _“button's hidden referenced name (visibility:hidden) with hidden aria-labelledby traversal falls back to aria-label”_ subtest at https://wpt.fyi/results/accname/name/comp_labelledby.html won’t pass unless when performing the “Return the accumulated text” substep of the “LabelledBy” substep in the accessible-name computation algorithm, I check to determine of the `total_accumulated_text` value is empty, and if so, then check for any `aria-label` attribute and use its value if found.

### Link to the version of the specification or documentation you were looking at at.

https://w3c.github.io/accname/#comp_labelledby

Does the issue exists in [the editors draft](https://w3c.github.io/aria/) (the editors draft is the most recent draft of the specification)? Yes

**Comment by accdc** [comment]
Hi,
I could just be tired, but I'm having trouble piecing together the issue.

Can you provide an HTML code example here and walk me through what the issue is?

Thanks,
Bryan

**Comment by MelSumner** [comment]
@sideshowbarker trying to resolve this issue. 

1. Can you confirm if you are still experiencing this problem?
2. If so, can you link to the test code (vs the test result) so we can inspect, or provide some markup/code samples here for us to look at and investigate? 

Thank you!

**Comment by sideshowbarker** [comment]
> 1. Can you confirm if you are still experiencing this problem?

Yes — the necessary steps are still missing from the spec.

> 2. If so, can you link to the test code (vs the test result) so we can inspect, or provide some markup/code samples here for us to look at and investigate?

The test case is at https://wpt.fyi/results/accname/name/comp_labelledby.html — and the two specific subtests are:

- “div group explicitly labelledby self and heading”
- “button's hidden referenced name (visibility:hidden) with hidden aria-labelledby traversal falls back to aria-label”

The relevant code for my implementation here:
 - https://github.com/LadybirdBrowser/ladybird/blob/f0917ea150fa1e75dd6910f8ec41e32b508942ea/Libraries/LibWeb/DOM/Node.cpp#L2528-L2551
- https://github.com/LadybirdBrowser/ladybird/blob/f0917ea150fa1e75dd6910f8ec41e32b508942ea/Libraries/LibWeb/DOM/Node.cpp#L2553-L2560
```c++
// ii. For each IDREF:
for (auto const& id_ref : id_list) {
    auto node = document.get_element_by_id(MUST(FlyString::from_utf8(id_ref)));
    if (!node)
        continue;
    // AD-HOC: The “For each IDREF” substep in the spec doesn’t seem to explicitly require the following
    // check for an aria-label value; but the “div group explicitly labelledby self and heading” subtest at
    // https://wpt.fyi/results/accname/name/comp_labelledby.html won’t pass unless we do this check.
    // https://github.com/w3c/aria/issues/2388
    if (target == NameOrDescription::Name && node->aria_label().has_value() && !node->aria_label()->is_empty() && !node->aria_label()->bytes_as_string_view().is_whitespace()) {
        total_accumulated_text.append(' ');
        total_accumulated_text.append(node->aria_label().value());
    }
    if (visited_nodes.contains(node->unique_id()))
        continue;

    // a. Set the current node to the node referenced by the IDREF.
    current_node = node;
    // b. Compute the text alternative of the current node beginning with step 2. Set the result to that text alternative.
    auto result = TRY(node->name_or_description(target, document, visited_nodes));
    // c. Append the result, with a space, to the accumulated text.
    total_accumulated_text.append(' ');
    total_accumulated_text.append(result);
}
```
```c++
// iii. Return the accumulated text.
// AD-HOC: This substep in the spec doesn’t seem to explicitly require the following check for an aria-label
// value; but the “button's hidden referenced name (visibility:hidden) with hidden aria-labelledby traversal
// falls back to aria-label” subtest at https://wpt.fyi/results/accname/name/comp_labelledby.html won’t pass
// unless we do this check.
// https://github.com/w3c/aria/issues/2388
if (total_accumulated_text.string_view().is_whitespace() && target == NameOrDescription::Name && element->aria_label().has_value() && !element->aria_label()->is_empty() && !element->aria_label()->bytes_as_string_view().is_whitespace())
    return element->aria_label().release_value();
return total_accumulated_text.to_string();
```

And what I’m stating in this issue is:

- If an implementation doesn’t do those two steps I’ve marked with `AD-HOC` comments in my code, then that implementation won’t be able to pass the test at https://wpt.fyi/results/accname/name/comp_labelledby.html.
- Therefore, the two steps I’ve described in the issue description should be added to the spec.
